### ISSUE TITLE
[Data] Implement dataset mixer for combining datasets in training

### ISSUE THREAD

**Issue by lewtun** [huggingface/trl#2112]
### Feature request

In the `alignment-handbook`, we implemented a ["dataset mixer"](https://github.com/huggingface/alignment-handbook/blob/7d711cd80dff17df78f8340df1b7616af0809407/src/alignment/data.py#L179) that allows one to easily combine datasets in varying proportions, provided they all share the same schema. 

It could be interesting to port this mixer to TRL, so that users can easily combine datasets during training. The only caveat I see is that to support the CLI training e.g. `trl sft ...` we'd need a data structure that is compatible because `dict` objects don't place nice with CLIs.

### Motivation

Advanced post-training typically combines different datasets / proportions. Supporting this in TRL would allow us to gradually deprecate the handbook in favour of using the lib directly.

### Your contribution

Open to discussion :)

**Comment by August-murr** [comment]
@lewtun 
Shouldn't this be a part of Datasets instead? 

Datasets already has [Interleaves](https://huggingface.co/docs/datasets/v3.0.1/en/process#interleave), which also mixes datasets in a similar way. It's not quite integrable as it is, but may be useful.

As for integrating it into the CLI, I think the best way to do it would be with a config file, like a JSON, something like this::

`trl sft --dataset_mixer  --mix_config mix_config.json`

mix_config.json:
```json
{
  "dataset_mixer": {
    "dataset_1": 0.4,
    "dataset_2": 0.3,
    "dataset_3": 0.2
  },
  "splits": ["train", "train", "test"],
  "configs": ["main", "math", "logic"],
  "columns_to_keep": ["text", "input", "text"],
  "shuffle": true
}
```
or a simpler JSON:
```json
[
  ["dataset_1", 0.4, "main", "train", "text"],
  ["dataset_2", 0.3, "math", "train", "input"],
  ["dataset_3", 0.2, "logic", "test", "text"],
]
```
if the parsing doesn't get too complicated. 

How does that sound?

**Comment by lewtun** [comment]
Yes good point about whether this should live in `datasets` or `trl`. Gently pinging @lhoestq for his thoughts on this. 

Context for Quentin: a common workflow in LLM fine-tuning is to mix datasets with mixed proportions (e.g. 30% of dataset together with 50% of dataset B). The `interleave_datasets()` method in `datasets` does something similar, but uses sampling and a stopping strategy which isn't quite the same thing we want here because we don't want to stop when one dataset is exhausted and we don't want to oversample either. One possibility would be to add a new stopping strategy like `none_exhausted` which samples the precise proportions (or less if there are not enough samples)

**Comment by lhoestq** [comment]
This works no ? And it feels easier to understand that a mixer (are the examples taken at random ? is the fraction representing the proportion of the original dataset or the resulting dataset ?)

```python
concatenate_datasets([
    dataset.select(range(int(frac * len(dataset))))
    for dataset, frac in zip(datasets, fracs)
])
```

**Comment by lewtun** [comment]
> This works no ? And it feels easier to understand than a mixer (are the examples taken at random ? is the fraction representing the proportion of the original dataset or the resulting dataset ?)
> 
> ```python
> concatenate_datasets([
>     dataset.select(range(int(frac * len(dataset))))
>     for dataset, frac in zip(datasets, fracs)
> ])
> ```

Yes, that is effectively what we did in the `alignment-handbook` [here](https://github.com/huggingface/alignment-handbook/blob/e057d7f45becb2bd4eef184c74a700132b2ec223/src/alignment/data.py#L179). In our case, the fractions represent the proportion per dataset and the samples are just taken from the first N rows. 

My question is more about whether it makes sense for `datasets` to have a utility method to load and combine _multiple_ datasets from the Hub? Here's how we did it in the handbook ([link](https://github.com/huggingface/alignment-handbook/blob/e057d7f45becb2bd4eef184c74a700132b2ec223/src/alignment/data.py#L125)) - happy to keep this logic in `trl` if it's too niche!

**Comment by lhoestq** [comment]
Since the mixing logic is kinda specific I don't think it should be in `datasets`. Though if there is something more general / explicit we can reconsider

**Comment by nZiben** [comment]
Sorry for the duplicate PR references here. I was testing the agent workflow and accidentally opened a couple of drafts. The duplicates are closed.

I opened #6171 with a small implementation for this issue: it adds a `fraction` field to `DatasetConfig`, so dataset mixtures can keep a deterministic proportion of each source dataset before concatenation.

**Comment by qgallouedec** [comment]
Hey @lewtun, this one's been open a while, but it's largely landed: the dataset mixer now lives in TRL as `DatasetMixtureConfig` + `get_dataset()`, wired into the sft/dpo/grpo/kto/reward/rloo CLIs via YAML.

The one remaining piece, per-dataset proportions, is addressed in #6199 (just opened), which adds a `fraction` field to `DatasetConfig` and keeps the first `int(fraction * len(dataset))` rows before concatenating.

**Comment by albertvillanova** [comment]
Mostly landed, thanks all. 

Before we close via #6199, I want to resurface one design point so we settle it on purpose rather than by default.

PR #6199 adds per-dataset fraction, implemented as the first `int(fraction * len(dataset))` rows of each source. That's precisely the alignment-handbook behavior, but it's also the answer to one of the two questions @lhoestq raised back in this comment:

> "is the fraction representing the proportion of the original dataset or the resulting dataset?"

PR #6199 chooses **proportion of the original**. 

But in my opinion, the downside is that the number isn't size-independent: 0.1 of a 100-row set and 0.1 of a 1T-row set mean very different contributions to the blend, so you can't actually specify a target mixture (e.g. 50/50) without knowing each source's size up front. For a "mixer," that's arguably the more natural thing to want.

There are three coherent semantics:
1. **Fraction of source** (what #6199 does): simple, composable, deterministic; but size-dependent and not a true mixture weight.
2. **Absolute row count**: size-independent, trivial to add, no normalization.
3. **Fraction of the final mixture** (normalized weights): the actual "blend" semantics; needs a base-size policy and truncates the largest source (the none_exhausted strategy mentioned earlier).

Worth a quick decision from @lewtun / @qgallouedec on which knobs we actually want to commit to in the public config surface.

Some ideas: keep per-source fraction as a primitive but document it unambiguously as downsampling a source (not a mixture weight), and consider adding absolute count as the size-independent knob; it's cheap and removes the footgun. True final-mixture proportions (3) can be a follow-up if there's demand.

**Comment by qgallouedec** [comment]
that a good point @albertvillanova. From a user's perspective, I find that specifying a proportion of the final dataset feels much more natural. I imagine that's generally what people want to specify.
